Feature/repository

Noor-Serry opened the pull request:
# Pull Request Template

## Description
Me And Sara : we implemented the dtos , mappers,localdatasource and setup the room database 
* Update *
Me and Nabil add the required function for our sub team in remote datasource 
---

## Changes Made
List the changes introduced in this pull request:

- room database implemention 
-daos
-dtos
- local mappers
---

## Screenshots (if applicable)
Include screenshots or GIFs to showcase the changes, especially if they impact the UI.


---

## Checklist
Please ensure the following tasks are completed:

- [x] My code follows the code style of this project
- [x] Changes have been tested manually and verified.
- [x] PR includes at most one single feature.

---

## Additional Comments
Add any additional information or context about the pull request here.

saraashraf11 committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/implement-search' into feature/repository

saraashraf11 committed:
feature: add repository implementation

saraashraf11 committed:
feature: add Room database

saraashraf11 committed:
feature: add Room entities for local data storage

saraashraf11 committed:
feature: add LocalTvShowDto

saraashraf11 committed:
feature: add Local Data Sources for Category, Movie, and Recent Search

saraashraf11 committed:
refactor: rename datasource directory to dataSource

Commit:
feature: Add country local data source and DTO

Commit:
feature: Add Kotlinx Datetime and enhance LocalCategoryDataSource

Commit:
refactor: Rename local package and update LocalMovieDataSource

Commit:
feature: add LocalTvShowDataSource and update LocalTvShowDto

Commit:
feature: Add SearchType enum and SearchTypeConverter

Commit:
refactor: rename dataSource package to datasource

saraashraf11 committed:
feature: add CategoryLocalMapper and MovieWithCategoryMapper

saraashraf11 committed:
refactor: rename CountryDto to LocalCountryDto

saraashraf11 committed:
feature: add country domain and repository

saraashraf11 committed:
refactor: rename SearchDto to LocalSearchDto

saraashraf11 committed:
feature: add TvShowLocalMapper

saraashraf11 committed:
feature: implement MovieLocalMapper

Noor-Serry committed:
Merge pull request #62 from Amsterdam-Team/feature/repositories-mapper

Feature/repositories mapper

Commit:
feature: implement local data sources and DAOs

This commit introduces implementations for local data sources and their corresponding Data Access Objects (DAOs)

Commit:
fix: remove Category entity from MovieWithCategories

Commit:
feature: add Country entity, CountryRepository, and local mappers

This commit introduces the following:

- `Country` data class in the `entity` module.
- `CountryRepository` interface in the `domain` module.
- `MovieWithCategoryMapper` to map between local DTOs and Movie entities.
- `MovieLocalMapper` for mapping between local Movie DTOs and Movie entities.
- `CountryLocalMapper` for mapping between local Country DTOs and Country entities.
- `CategoryLocalMapper` for mapping between local Category DTOs and Category entities.
- `TvShowLocalMapper` for mapping between local TvShow DTOs and TvShow entities.
- Renamed `SearchDto` to `LocalSearchDto`.
- Renamed `CountryDto` to `LocalCountryDto`.
- Formatted `Category` and `TvShow` data classes.

Commit:
Merge branch 'develop' of https://github.com/Amsterdam-Team/Aflami into feature/repository

Commit:
refactor: rename DTOs and update DAO interfaces

This commit renames several DTO classes to include "Local" as a prefix, improving clarity and consistency within the local datasource module.

Specifically, the following DTOs were renamed:
- `CountryDto` to `LocalCountryDto`
- `SearchDto` to `LocalSearchDto`

The corresponding DAO interfaces and their implementations (`CountryDao`, `LocalMovieDao`, `RecentSearchDao`, `TvShowDao`, `CountryDataSourceImpl`, `LocalMovieDataSourceImpl`, `RecentSearchDataSourceImpl`, `LocalTvShowDataSourceImpl`) have been updated to use the new DTO names.

Additionally, the `LocalRecentSearchDataSource` interface has been simplified by removing Room-specific annotations, as it's an interface in the repository layer.

The `LocalTvShowDataSourceImpl` and `TvShowDao` were updated to handle the relationship between `LocalTvShowDto` and search keywords by introducing `LocalTvShowWithSearchDto` and methods to insert these mappings.

Commit:
refactor : rename doa and add export schema to false

Commit:
refactor : rename doa and add export schema to false

Commit:
Merge remote-tracking branch 'origin/feature/repository' into feature/repository

Commit:
Merge remote-tracking branch 'origin/feature/repository' into feature/repository

Commit:
Merge remote-tracking branch 'origin/feature/repository' into feature/repository

Commit:
Merge remote-tracking branch 'origin/feature/repository' into feature/repository

Pixelh97 commented on ui/src/main/java/com/example/ui/screens/SearchByCriteria.kt:
```diff
@@ -0,0 +1,70 @@
+import com.example.designsystem.utils.ThemeAndLocalePreviews
+
+@Composable
+fun SearchByCriteria(
```
The screen name should contain a `Screen` keyword postfix for readability

Pixelh97 commented on ui/src/main/java/com/example/ui/screens/SearchByCriteria.kt:
```diff
@@ -0,0 +1,70 @@
+    Column {
+        DefaultAppBar(
+            title = "World Tour",
+            showNavigateBackButton = true,
+        )
```
Handling the navigation click is required here, even if it will have an empty function until the logic of navigation is done

Pixelh97 commented on ui/src/main/java/com/example/ui/screens/SearchByCriteria.kt:
```diff
@@ -0,0 +1,70 @@
+        ) {
+            TextField(
+                "",
+                hintText = stringResource(R.string.country_name_hint),
+            )
```
Also, handing the `onChangeValue` is required to be ready for the viewModel when it's done

Pixelh97 commented on ui/src/main/java/com/example/ui/screens/SearchByCriteria.kt:
Final note, any screen should have 2 composables, one holding the states and the other is for the components themselves without any state, so even if you don't have the state right now, you should leave a place for it.

Commit:
Merge remote-tracking branch 'origin/feature/repository' into feature/repository

Commit:
Merge remote-tracking branch 'origin/feature/repository' into feature/repository

Commit:
Merge remote-tracking branch 'origin/feature/repository' into feature/repository

Commit:
Refactor: update package name for Room DAOs

The package name for Room Data Access Objects (DAOs) has been changed from `com.example.localdatasource.roomDatabase.daos` to `com.example.localdatasource.roomDataBase.daos`.

This change affects the following files:
- `CountryDataSourceImpl.kt`
- `LocalCategoryDataSourceImpl.kt`
- `TvShowDao.kt`
- `MovieDao.kt`
- `RecentSearchDao.kt`
- `CategoryDao.kt`
- `CountryDao.kt`
- `AflamiDatabase.kt`
- `RecentSearchDataSourceImpl.kt`
- `LocalMovieDataSourceImpl.kt`
- `LocalTvShowDataSourceImpl.kt`

Additionally, a new function `getSearchInfo` has been added to `LocalRecentSearchDataSource.kt` and `RecentSearchDao.kt` to retrieve search information by keyword.

Commit:
feature: implement remote movie data source

This commit introduces the functionality to fetch movies from a remote API.

Key changes include:
- Added `RemoteMovieDatasource` interface in the repository module and its implementation `RemoteMovieDatasourceImpl` in the remoteDatasource module.
- `RemoteMovieDatasourceImpl` uses Ktor client to fetch movies based on keyword, rating, and category ID.
- Introduced DTOs (`RemoteMovieResponse`, `RemoteMovieItemDto`) for parsing the API response.
- Added `RemoteMovieMapper` to map remote DTOs to domain `Movie` objects.
- Added kotlinx.serialization plugin and dependency to the repository module for JSON parsing.
- Organized mappers into `local` and `remote` subpackages within the repository module.
- Defined a `BASE_URL` constant for the remote API.

nabilBouzineDev committed:
refactor: add TV Show data fetching capabilities

eslam-99 reviewed (commented):
I think the merge should be into `feature/implement-search` not `develop`

eslam-99 commented on domain/src/main/java/com/example/domain/repository/CountryRepository.kt:
```diff
@@ -0,0 +1,7 @@
+import com.example.entity.Country
+
+interface CountryRepository {
+    suspend fun getSuggestedCountries(): List<Country>
```
As the filtration will be done in domain so the repo will just use `getAllCountries` instead of `getSuggestedCountries` and then will be filtered in usecase

nabilBouzineDev replied:
Done ✅

eslam-99 commented on domain/src/main/java/com/example/domain/repository/MovieRepository.kt:
```diff
@@ -0,0 +1,9 @@
+
+interface MovieRepository {
+    suspend fun getMoviesByKeyword(keyword: String): List<Movie>
+    suspend fun getMoviesByActor(actorName: String): List<Movie>
```
If there will be auto complete in actor names as countries, so we will pass `actorId` here, please check with @marah222 as she asked on slack

nabilBouzineDev replied:
Ok we discussed on the meeting and we will implement autocomplete based on the returned type

eslam-99 commented on domain/src/main/java/com/example/domain/repository/RecentSearchRepository.kt:
```diff
@@ -0,0 +1,8 @@
+
+interface RecentSearchRepository {
+    suspend fun insertOrReplaceRecentSearch(searchKeyword: String)
+    suspend fun getRecentSearch(): List<String>
```
Will be `getRecentSearches` as it's plural and returns a list

nabilBouzineDev replied:
Done ✅

nabilBouzineDev committed:
Merge branch 'feature/implement-search' of github.com:Amsterdam-Team/Aflami into feature/repository

# Conflicts:
#	domain/build.gradle.kts
#	entity/src/main/java/com/example/entity/Country.kt
#	ui/src/main/java/com/example/ui/screens/SearchByCountryScreen.kt

nabilBouzineDev committed:
refactor: rename `getSuggestedCountries` to `getAllCountries`

nabilBouzineDev committed:
refactor: rename `getRecentSearch` to `getAllRecentSearches`

nabilBouzineDev commented:
> I think the merge should be into `feature/implement-search` not `develop`

Ok @eslam-99  I've changed the base branch to be implement search
